mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendWelcomeEmailAction.php

```diff
@@ -0,0 +1,139 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Automation\Integrations\MailPoet\Actions;
+
+use MailPoet\Automation\Engine\Workflows\Action;
+use MailPoet\Automation\Engine\Workflows\Step;
+use MailPoet\Automation\Engine\Workflows\Workflow;
+use MailPoet\Automation\Engine\Workflows\WorkflowRun;
+use MailPoet\Automation\Integrations\MailPoet\Subjects\SegmentSubject;
+use MailPoet\Automation\Integrations\MailPoet\Subjects\SubscriberSubject;
+use MailPoet\Entities\NewsletterEntity;
+use MailPoet\Entities\SubscriberEntity;
+use MailPoet\InvalidStateException;
+use MailPoet\Newsletter\NewslettersRepository;
+use MailPoet\Newsletter\Scheduler\WelcomeScheduler;
+use MailPoet\Newsletter\Sending\ScheduledTasksRepository;
+use MailPoet\Subscribers\SubscriberSegmentRepository;
+use MailPoet\Subscribers\SubscribersRepository;
+
+class SendWelcomeEmailAction implements Action {
+  /** @var NewslettersRepository */
+  private $newslettersRepository;
+
+  /** @var ScheduledTasksRepository */
+  private $scheduledTasksRepository;
+
+  /** @var SubscriberSegmentRepository */
+  private $subscriberSegmentRepository;
+
+  /** @var SubscribersRepository */
+  private $subscribersRepository;
+
+  /** @var WelcomeScheduler */
+  private $welcomeScheduler;
+
+  public function __construct(
+    NewslettersRepository $newslettersRepository,
+    ScheduledTasksRepository $scheduledTasksRepository,
+    SubscriberSegmentRepository $subscriberSegmentRepository,
+    SubscribersRepository $subscribersRepository,
+    WelcomeScheduler $welcomeScheduler
+  ) {
+    $this->newslettersRepository = $newslettersRepository;
+    $this->scheduledTasksRepository = $scheduledTasksRepository;
+    $this->subscriberSegmentRepository = $subscriberSegmentRepository;
+    $this->subscribersRepository = $subscribersRepository;
+    $this->welcomeScheduler = $welcomeScheduler;
+  }
+
+  public function getKey(): string {
+    return 'mailpoet:send-welcome-email';
+  }
+
+  public function getName(): string {
+    return __('Send Welcome Email', 'mailpoet');
+  }
+
+  public function isValid(array $subjects, Step $step, Workflow $workflow): bool {
+    try {
+      $this->getWelcomeEmailForStep($step);
+    } catch (InvalidStateException $exception) {
+      return false;
+    }
+    
+    $segmentSubject = $subjects['mailpoet:segment'] ?? null;
+    $subscriberSubject = $subjects['mailpoet:subscriber'] ?? null;
+
+    return $segmentSubject instanceof SegmentSubject && $subscriberSubject instanceof SubscriberSubject;
+  }
+
+  private function getWelcomeEmailForStep(Step $step): NewsletterEntity {
+    if (!isset($step->getArgs()['welcomeEmailId'])) {
```

Comment:
Thank you for pointing out these things! I always appreciate when people point out "nitpicky" things like this, and I don't even like calling it nitpicky because I think they're important in their own way.